src/doppler-types.ts

```diff
@@ -64,39 +64,210 @@ export interface PrepaidPack {
+  ) as SubscribersLimitedPlan[];
+
+const getFreePlans = (planList: Plan[]): FreePlan[] =>
+  planList.filter((x) => x.type == 'free') as FreePlan[];
+
+const getAgencyPlans = (planList: Plan[]): AgencyPlan[] =>
+  planList.filter((x) => x.type == 'agency') as AgencyPlan[];
+
+const getStandardPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet == 'standard') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getPlusPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet == 'plus') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getFreePathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [FreePath] => {
+  var plans = getFreePlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'free',
+      actual: userPlan.featureSet == 'free',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getAgencyPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [AgenciesPath] => {
+  var plans = getAgencyPlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'agencies',
+      actual: userPlan.featureSet == 'agency',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getStandardPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [StandardPath] => {
+  var plans = getStandardPlans(planList).sort((x) => x.fee);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'standard',
+      actual: userPlan.featureSet == 'standard',
+      minimumFee: cheapestPlan.fee,
+      deadEnd: plans.length == 1 && cheapestPlan == userPlan,
+    },
+  ];
+};
+
+const getPlusPathOrEmpty = (userPlan: Plan, planList: Plan[]): PlusPath[] => {
+  var plans = getPlusPlans(planList).sort((x) => x.fee);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'plus',
+      actual: userPlan.featureSet == 'plus',
+      minimumFee: cheapestPlan.fee,
+      deadEnd: plans.length == 1 && cheapestPlan == userPlan,
+    },
+  ];
+};
+
+const _agencyPlan: AgencyPlan = {
+  type: 'agency',
+  featureSet: 'agency',
+};
+
+export const getPaths = (userPlan: Plan, planList: Plan[]): Path[] => {
+  // Assuming that planList does not contains 'free' and neither 'agency'
+  const potentialUpgradePlans =
+    userPlan.type == 'free'
+      ? planList
+      : userPlan.type == 'prepaid'
+      ? [
+          ...getPrepaidPacks(planList),
+          ...getUpgradeMonthlyPlans(planList),
+          ...getUpgradeSubscribersPlans(planList),
+        ]
+      : userPlan.type == 'monthly-deliveries'
```

Comment:
Here potential upgrades for 'monthly-deliveries' should include 'subscribers' plans. But the trick here is we must not show plans with less subscribers than what the user already has. So for a 'monthly-deliveries' plan with 2300 subscribers in the account we need to show plans above 2300 subscribers only.  
For 'subscribers' plan there is no issue, I would compare by fee, show monthly plans with more fee.

Reply:
Wow, it is a really complex business rule. Are we supporting that right now? 

I think that by the moment we can restrict that path, it could be easily added in the future.

Reply:
Right now you can do this kind of upgrade manually by admin. In app cannot be done yet.

Reply:
Well, in that case, we do not need to implement it now, but your proposal looks clean and correct.

Reply:
I will leave only the 'subscribers' -> 'monthly-deliveries' possibility, that is the easy way. 
The other way around I will have to give it more thought...